src/lib/tools.ts

```diff
@@ -1,123 +1,199 @@
+    x2: number,
+    y2: number
+  ): number[] => {
+    const x = x2 - ox;
+    const y = y2 - oy;
+    const side = vx * y - vy * x;
+    const sq = vx * vx + vy * vy;
+    return [
+      Math.abs(side) / sq,
+      ox + x + (vy * side) / sq,
+      oy + y - (vx * side) / sq,
+    ];
+  };
+}
+
+export class ToolHandler {
+  // Add type marker `boolean` so it can be overridden; otherwise takes type `false`
+  /**
+   * Make sure that no extending classes except BrushHandler set this to true;
+   * the value of this property is used as a type guard.
+   */
+  readonly isBrush: boolean = false;
+  /**
+   * Make sure that no extending classes except DrawingToolHandler set this to true;
+   * the value of this property is used as a type guard.
+   */
+  readonly isDrawing: boolean = false;
+  /**
+   * Make sure that no extending classes except DrawingToolHandler set this to true;
+   * the value of this property is used as a type guard.
+   */
+  readonly requiresBase: boolean = false;
+
+  resize: (
     object: fabric.Object,
     x2: number,
     y2: number,
     strict: boolean
-  ) => Promise<fabric.Object>;
+  ) => fabric.Object | Promise<fabric.Object>;
+
+  /**
+   * Set externally from activate().
+   * Set internally with setActive();
+   * @private
+   */
+  private active: boolean;
+
+  constructor(
+    protected baseCanvas: Page,
+    protected history: HistoryHandler,
+    protected clipboard: ClipboardHandler
+  ) {}
+
+  /**
+   * @return Whether the activation was successful.
+   * Maybe want to throw error instead of return boolean.
+   */
+  activate: () => boolean | Promise<boolean> = () => this.setActive(true);
+
+  /**
+   * Not allowed to fail
+   */
+  deactivate: () => void = () => this.setActive(false);
+
+  /**
+   * get this.active
+   */
+  isActive = () => this.active;
+
+  /**
+   * set this.active active
+   */
+  protected setActive = (active: boolean) => (this.active = active);
+}
 
-  setBrush?: (
-    brush: fabric.BaseBrush,
-    options: fabric.IObjectOptions
-  ) => Promise<void>;
+export abstract class DrawingToolHandler extends ToolHandler {
+  isDrawing = true;
+  abstract draw: (
+    x,
+    y,
+    options,
+    x2?,
+    y2?
+  ) => fabric.Object | Promise<fabric.Object>;
 }
 
-export class MoveHandler implements ToolHandler {
-  tool: Tool = Tool.Move;
-  isBrush = false;
+export function isDrawing(tool: ToolHandler): tool is DrawingToolHandler {
+  return tool.isDrawing;
 }
 
-export class PenHandler implements ToolHandler {
-  tool: Tool = Tool.Pen;
+export abstract class BrushHandler extends ToolHandler {
   isBrush = true;
 
-  setBrush = async (
+  /**
+   * Handle the pathCreated event
+   */
+  abstract pathCreated: (e: any) => void | Promise<void> = () => {};
+
+  abstract setBrush: (
     brush: fabric.BaseBrush,
     options: fabric.IObjectOptions
-  ): Promise<void> => {
+  ) => void | Promise<void> = () => {};
+}
+
+export function isBrush(tool: ToolHandler): tool is BrushHandler {
```

Comment:
why does this exist? why don't we just check tool.isbrush? same re tool.isdrawing

Reply:
i think it makes more sense that "isBrush" is a property of the tool, rather than the module?

Reply:
I can double check, but the only reason I'd add this is because without it, typescript doesn't have sufficient information to distinguish between tool and non-tool elements. This is important because you call specific methods, like setBrush() if and only if the tool is a brush.

Re: your second comment: this function is a typescript type guard. It's a javascript function that returns a special boolean: the return value (of type `tool is BrushHandler`) tells TS whether `tool` can be assigned the type `BrushHandler`.

This guard function itself is a property of the module, but it queries the tool's own `isBrush` property. i.e., in another module, you do
```ts
declare const MyTool: ToolHandler; //  just some tool instance that exists because of something

if(isBrush(MyTool)) tool.setBrush(...)
```

Reply:
Thank you for bringing this up actually. In dde7fcefd02e01dac166206e6615eeab6d8b1877, I have moved these type guards "type predicates" to be members of the base class, which I think makes it a bit more readable. Typescript unfortunately doesn't let you use properties themselves as type guards, so we must store a private property instead.